torchaudio/prototype/io/streamer.py

```diff
@@ -0,0 +1,486 @@
+
+    @property
+    def num_out_streams(self):
+        """Number of output streams configured by client code.
+
+        :type: int
+        """
+        return torch.ops.torchaudio.ffmpeg_streamer_num_out_streams(self._s)
+
+    @property
+    def default_audio_stream(self):
+        """The index of default audio stream. ``None`` if there is no audio stream
+
+        :type: Optional[int]
+        """
+        return self._i_audio
+
+    @property
+    def default_video_stream(self):
+        """The index of default video stream. ``None`` if there is no video stream
+
+        :type: Optional[int]
+        """
+        return self._i_video
+
+    def get_src_stream_info(self, i: int) -> torchaudio.prototype.io.SourceStream:
+        """Get the metadata of source stream
+
+        Args:
+            i (int): Stream index.
+        Returns:
+            SourceStream
+        """
+        return _parse_si(torch.ops.torchaudio.ffmpeg_streamer_get_src_stream_info(self._s, i))
+
+    def get_out_stream_info(self, i: int) -> torchaudio.prototype.io.OutputStream:
+        """Get the metadata of output stream
+
+        Args:
+            i (int): Stream index.
+        Returns:
+            OutputStream
+        """
+        return _parse_oi(torch.ops.torchaudio.ffmpeg_streamer_get_out_stream_info(self._s, i))
+
+    def seek(self, timestamp: float):
+        """Seek the stream to the given timestamp [second]
+
+        Args:
+            timestamp (float): Target time in second.
+        """
+        torch.ops.torchaudio.ffmpeg_streamer_seek(self._s, timestamp)
+
+    def add_basic_audio_stream(
+        self,
+        frames_per_chunk: int,
+        buffer_chunk_size: int = 3,
+        stream_index: Optional[int] = None,
+        sample_rate: Optional[int] = None,
+        dtype: Optional[torch.dtype] = torch.float32,
+    ):
+        """Add output audio stream
+
+        Args:
+            frames_per_chunk (int): A unit number of frames returned by Streamer as a chunk.
+                If the source stream is exhausted before enough frames are buffered,
+                then the chunk is returned as-is.
+
+            buffer_chunk_size (int, optional): Internal buffer size.
+                When this many chunks are created, but
+                client code does not pop the chunk, if a new frame comes in, the old
+                chunk is dropped.
+
+            stream_index (int or None, optional): The source audio stream index.
+                If omitted, :py:attr:`default_audio_stream` is used.
+
+            sample_rate (int or None, optional): If provided, resample the audio.
+
+            dtype (torch.dtype, optional): If not ``None``, change the output sample precision.
+                If floating point, then the sample value range is
+                `[-1, 1]`.
+        """
+        i = self.default_audio_stream if stream_index is None else stream_index
+        torch.ops.torchaudio.ffmpeg_streamer_add_basic_audio_stream(
```

Comment:
> if one output stream advances more quickly than the others, e.g. its `frames_per_chunk`-to-`sample_rate` ratio is higher, are the other output streams guaranteed to be not affected?

Under the circumstance the value of `buffer_chunk_size` is large enough so that no chunk is dropped during the streaming, then the other output streams are not affected.

The media source with multiple streams, like mp4 is composed of the sequence of containers corresponding to the different streams. If one output stream is configured to have higher `frames_per_chunk / sample_rate` value, then when the `Streamer` attempts to fill all output stream buffers, even after the other output streams have enough frames to consist the first chunk, this output stream with higher `frames_per_chunk / sample_rate` value can be still missing frames for the first chunk. Then the `Streamer` has to parse more frames to fill the buffer of this stream. When the `Streamer` parse more frames, inevitably it will also parse the frames of the other streams as well.

![Untitled drawing](https://user-images.githubusercontent.com/855818/151487352-3d05b9d9-5865-4a4c-8fad-aec7da584113.png)

Reply:
i see. i'm not very familiar with ffmpeg — would most users generally know that chunks would be dropped if output streams are configured differently? wondering if it'd make sense to, say, surface a warning

Reply:
Okay, added warnings [here](https://github.com/pytorch/audio/pull/2164/commits/5f531fa71d35e85ac8d37057d28e340b68591222).